File: Homework1/Codes/22202709_Ozongun_HalilArda.kt
fun printArray(arr: Any?) {
    when (arr) {
        is IntArray -> for (item in arr) print("$item ")
        is Array<*> -> for (item in arr) print("$item ")
        is Iterable<*> -> for (item in arr) print("$item ")
        }
    println()
}

fun main(){
    println("1. Legal types for subscripts:")
    val arr1 = arrayOf(1, 2, 3, 4, 5)
    println(arr1[0])
    // println(arr1[0.5]) Compile time error: argument type mismatch
    // println(arr1["3"]) Compile time error: argument type mismatch
    // println(arr1[true])  Compile time error: argument type mismatch
    // println(arr1[0..2]) argument type mismatch: actual type is 'IntRange', but 'Int' was expected.
    /*
    var ind : Long = 3
    println(arr1[ind]) // Compile time error: argument type mismatch: actual type is 'Long', but 'Int' was expected.
    */
    /*
    var ind : Short = 3
    println(arr1[ind]) // Compile time error: argument type mismatch: actual type is 'Short', but 'Int' was expected.
    */
    

    println("\n2. Range checking in subscripting expressions")
    // println(arr1[6]) // Runtime error: ArrayIndexOutOfBoundsException
    // println(arr1[-1]) // Runtime error: ArrayIndexOutOfBoundsException
    val num = 68
    // println(arr1[num]) // Runtime error: ArrayIndexOutOfBoundsException.


    println("\n3. Ragged multidimensional arrays")
    val raggedArr = arrayOf(arrayOf(1, 2, 3), arrayOf(4, 5), arrayOf(6, 7, 8, 9))
    for (i in raggedArr.indices) {
        printArray(raggedArr[i]) 
    }
    println("Type of raggedArr: ${raggedArr::class.simpleName}") // Array<Array<Int>>
    println("Type of raggedArr[0]: ${raggedArr[0]::class.simpleName}") // Array<Int>
    println("Type of raggedArr[0][0]: ${raggedArr[0][0]::class.simpleName}") // Int


    

    println ("\n4. Initialization of arrays")

    val arr2 = arrayOf(1, 2, 3, 4, 5, "Ataturk", 3.14, true) // mix type
    printArray(arr2) 

    val intArr = IntArray(5) // size 5, 00000
    printArray(intArr)

    val lambdaArr = Array(5) { it * 2 }
    printArray(lambdaArr) // 0 2 4 6 8

    val arr4 = arrayOfNulls<Int>(5)
    printArray(arr4) 
    println(arr4::class.simpleName)

    val arr5 = Array<Int?>(5) { null }
    printArray(arr5)



    println("\n5. Array slices")

    val arr6 = arrayOf(1, 2, 3, 4, 5)
    val slice = arr6.slice(1..3) // (begin, end) inclusive
    println("Kind of slice: ${slice::class.simpleName}") 
    printArray(slice)
    

    val slice3 = arr6.slice(0..2 step 2) // step 2
    println("Kind of slice: ${slice3::class.simpleName}") 
    printArray(slice3)

    // To achive array when sliced:
    val coppied = arr6.copyOfRange(1, 3) // (begin, end) exclusive
    println("Kind of slice: ${coppied::class.simpleName}") 
    printArray(coppied)




    println("\n6.Allowed Operations")
    val a = arrayOf(1, 2, 3)
    val b = arrayOf(4, 5, 6)
    val c = arrayOf(1, 2, 3)
    print("a: ")
    printArray(a)
    print("b: ")
    printArray(b)
    print("c: ")
    printArray(c)

    println("Compare contents:") // Checks if the two specified arrays are structurally equal to one another.
    println("a contentEquals b: ${a contentEquals b}") 
    println("a contentEquals c: ${a contentEquals c}") 

    println("Structural equality:") // two objects have the same content or structure, but for arrays: assigned variables point to the same object.
    println("a == b: ${a == b}") 
    println("a == c: ${a == c}")

    println("Referential equality:") // two references pointing to the same object
    println("a === b: ${a === b}") 
    println("a === c: ${a === c}")

    val d = a
    println("a == d: ${a == d}")
    println("a === d: ${a === d}") 

    println("Concatenation:")
    val concat = a + b
    printArray(concat)

    println("Contains:")
    println("1 in a: ${1 in a}")
    println("10 in a: ${10 in a}") 
}

File: Homework1/Codes/22202709_Ozongun_HalilArda.dart
import 'package:collection/collection.dart';

void main(){
  print("1: Legal types for subscripts");
  List<int> arr1 = [1, 2, 3, 4, 5];
  print(arr1[3]);
  // print(arr1[3.0]);    Compile time error
  // print(arr1["1"]);    Compile time error
  // print(arr1[true]);   Compile time error         
  /*double floatIndex = 2.8;
  print(arr1[floatIndex]); // Compile time error */


  print("\n\n2: Range checking in subscripting expressions:");
  // print(arr1[10]); // Runtime error 
  // print(arr1[-1]); // Runtime error
  int index = 4324;
  // print(arr1[index]);  // Runtime error



  print("\n\n3: Ragged multidimensional arrays:");
  List<List<int>> ragged = [
    [1, 2, 3],
    [4, 5],
    [6, 7, 8, 9]
  ];

  print(ragged);
  print(ragged is List<List<int>>);
  print(ragged[0] is List<int>);       



  print("\n\n4: Initialization of arrays:");
  var init1 = [1, "ataturk", 3, true];
  print("init1: $init1");
  var init2 = List.filled(5, null, growable: false);
  print("init2: $init2"); 
  var init3 = List<int>.generate(8, (i) => i *  i); 
  print("init3: $init3");
  var init4 = List.of([5]);
  print("init4: $init4");
  const init5 = [1, 2, 3, 4, 5];
  print("init5: $init5");



  print("\n\n5: Array Slice:");
  arr1 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
  var slice1 = arr1.sublist(2, 5); // (start,end)
  print("slice1 : $slice1"); 
  var slice2 = arr1.sublist(3); // start
  print("slice2 : $slice2");
  // First N elements
  var slice3 = arr1.take(4).toList();
  print("slice3: $slice3");
  // skip first n elements
  var slice4 = arr1.skip(5).toList();
  print("slice4: $slice4");
  // Filtering
  var slice5 = arr1.where((x) => x > 5).toList();
  print("slice5: $slice5");



  print("\n\n6: Allowed operations:");

  var a = [1, 2, 3];
  var b = [4, 5, 6];
  var c = [1, 2, 3];
  
  print("a: $a, b: $b, c: $c");

  print("a == c: ${a == c}"); // false, different objects
  print("identical(a, c): ${identical(a, c)}"); // false, different objects

  var d = a; 
  print("a == d: ${a == d}"); // true, same object
  print("identical(a, d): ${identical(a, d)}"); // true, same object

  Function eq = const ListEquality().equals;
  print("ListEquality: ${eq(a, c)}"); 

  // concatenation
  var ab = a + b;
  print("Concatenation: $ab");

  var concat = [...a, ...b];
  print("Concatenation: $concat"); 

  // Contains
  print("2 in a: ${a.contains(2)}");
  print("10 in a: ${a.contains(10)}");

}